# Day 11, part 1

Each fuel cell has a coordinate ranging from 1 to 300 in both the X (horizontal) and Y (vertical) direction. In X,Y notation, the top-left cell is 1,1, and the top-right cell is 300,1.

The interface lets you select any 3x3 square of fuel cells. To increase your chances of getting to your destination, you decide to choose the 3x3 square with the largest total power.

The power level in a given fuel cell can be found through the following process:

- Find the fuel cell's rack ID, which is its X coordinate plus 10.
- Begin with a power level of the rack ID times the Y coordinate.
- Increase the power level by the value of the grid serial number (your puzzle input).
- Set the power level to itself multiplied by the rack ID.
- Keep only the hundreds digit of the power level (so 12345 becomes 3; numbers with no hundreds digit become 0).
- Subtract 5 from the power level.

For example, to find the power level of the fuel cell at 3,5 in a grid with serial number 8:

- The rack ID is 3 + 10 = 13.
- The power level starts at 13 * 5 = 65.
- Adding the serial number produces 65 + 8 = 73.
- Multiplying by the rack ID produces 73 * 13 = 949.
- The hundreds digit of 949 is 9.
- Subtracting 5 produces 9 - 5 = 4.

So, the power level of this fuel cell is 4.

What is the X,Y coordinate of the top-left fuel cell of the 3x3 square with the largest total power?

Your puzzle input is 7689.

** correct answer: 20 37 **

In [1]:
from collections import defaultdict
import numpy as np

# puzzle input = serial number
serial = 7689
grid_size = 300

def f(x, y, s):
    rack_id = x + 10 # rack id = x + 10 (add 1 as grid offset)
    val = rack_id * y  # power level = rack id * y (add 1 as grid offset)
    val += s # add serial number
    val *= rack_id # multiply by rack ID
    val = val // 100 % 10 # get 100 digit
    val -= 5 # take 5 off for final result
    return val

# function that summarizes slices of the table with size "z"
def f_grid(x, y, z, grid):
    return np.sum(grid[x:x+z, y:y+z])
    

# create the initial grid using the creation formula
grid = []
for x in range(1, grid_size + 1):
    grid_row = []
    for y in range(1, grid_size + 1):
        grid_row.append(f(x, y, serial))
    grid.append(grid_row)
    
# convert the grid to a numpy array for easy sums across slices (for part 2)
grid_np = np.array(grid)

# create a dictionary that stores values for each subgrid
# keys are x, y coordinates (top left corner) and size (z)
sum_grid = defaultdict(int)
for z in range(1, grid_size + 1):
    print('z = %d' % (z))
    for x in range(1, grid_size + 1 - z):
        for y in range(1, grid_size + 1 - z):
            sum_grid[(x, y, z)] = f_grid(x, y, z, grid_np)
    
def print_grid(x, y, z, grid):
    for j in range(y, y+z):
        output = ''
        for i in range(x, x+z):
            output = output + str(grid[i][j]) + ' '
        print(output)
        
#print(print_grid(32, 44, grid))
#print(print_grid(89, 268, 16, grid))


z = 1
z = 2
z = 3
z = 4
z = 5
z = 6
z = 7
z = 8
z = 9
z = 10
z = 11
z = 12
z = 13
z = 14
z = 15
z = 16
z = 17
z = 18
z = 19
z = 20
z = 21
z = 22
z = 23
z = 24
z = 25
z = 26
z = 27
z = 28
z = 29
z = 30
z = 31
z = 32
z = 33
z = 34
z = 35
z = 36
z = 37
z = 38
z = 39
z = 40
z = 41
z = 42
z = 43
z = 44
z = 45
z = 46
z = 47
z = 48
z = 49
z = 50
z = 51
z = 52
z = 53
z = 54
z = 55
z = 56
z = 57
z = 58
z = 59
z = 60
z = 61
z = 62
z = 63
z = 64
z = 65
z = 66
z = 67
z = 68
z = 69
z = 70
z = 71
z = 72
z = 73
z = 74
z = 75
z = 76
z = 77
z = 78
z = 79
z = 80
z = 81
z = 82
z = 83
z = 84
z = 85
z = 86
z = 87
z = 88
z = 89
z = 90
z = 91
z = 92
z = 93
z = 94
z = 95
z = 96
z = 97
z = 98
z = 99
z = 100
z = 101
z = 102
z = 103
z = 104
z = 105
z = 106
z = 107
z = 108
z = 109
z = 110
z = 111
z = 112
z = 113
z = 114
z = 115
z = 116
z = 117
z = 118
z = 119
z = 120
z = 121
z = 122
z = 123
z = 124
z = 125
z = 126
z = 127
z = 128
z = 129
z = 130
z = 131
z = 132
z = 133
z = 134
z = 135
z = 136
z = 137
z = 138
z = 

In [3]:
print(f_grid(89, 168, 15, grid_np))

103


In [4]:
max_x, max_y, max_z = max(sum_grid, key = sum_grid.get)
print(max_x + 1, max_y + 1, max_z)
        

90 169 15


# part 2


correct answer: 90 169 15

# Part 1 and part 2 - better solution

Using partial sum table

In [37]:
from collections import defaultdict

# puzzle input = serial number
serial = 7689
grid_size = 300

def f(x, y, s):
    rack_id = x + 1 + 10 # rack id = x + 10 (add 1 as grid offset)
    val = rack_id * (y + 1)  # power level = rack id * y (add 1 as grid offset)
    val += s # add serial number
    val *= rack_id # multiply by rack ID
    val = val // 100 % 10 # get 100 digit
    val -= 5 # take 5 off for final result
    return val

# calculate grids
grid = [[f(x, y, serial) for y in range(300)] for x in range(300)]

# create the partial table list
part_sum = [[0 for _ in range(300)] for _ in range(300)]
# function that returns partial sum value (0 if outside of bounds)
def get_part_sum(x: int, y: int) -> int:
    if x < 0 or y < 0:
        return 0
    else:
        return part_sum[x][y]
    
# create partial sums
# A B C
# D E F
# G H I
# partial sum for I =
# H + F + f(I) - E
# (x-1, y) + (x, y-1) + f(I) - (x-1, y-1)
for x in range(300):
    for y in range(300):
        part_sum[x][y] = get_part_sum(x-1, y) + get_part_sum(x, y-1) + grid[x][y] - get_part_sum(x-1, y-1)
    
    
# generate grid dictionary
grid_dict = defaultdict(int)
# calculate squares
# A B C D
# E F G H
# I J K L
# M N O P
# K+1 (KLOP) = P - N - H + F
# F+2 (FGHJKLNOP) = P (x+2, y+2) - D (x+2, y-1) - M (x-1, y+2) + A (x-1, y-1)
# add 1 to s (size) since we go from grid size 1 (just one element) to grid size 2 (4 elements)
for s in range(0, 300):
    for x in range(300 - s):
        for y in range(300 - s):
            grid_dict[(x, y, s+1)] = get_part_sum(x+s, y+s) - get_part_sum(x+s, y-1) - get_part_sum(x-1, y+s) + get_part_sum(x-1, y-1) 


103


In [39]:
max_x, max_y, max_z = max(grid_dict, key = grid_dict.get)
print(max_x + 1, max_y + 1, max_z)


90 169 15
